---
title: "Article data analysis"
author: "HÃ¥kon Kaspersen"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
always_allow_html: yes
output:
    html_document:
      code_folding: hide
      df_print: paged
      highlight: kate
      number_sections: yes
      theme: flatly
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: yes
    pdf_document:
      toc: yes
      toc_depth: '3'
---

# Background

## Aims

- Identify potential differences in resistance mechanisms between wild- and production animals.

- Identify potential clones which are species specific.

- Characterize resistance mechanisms in QREC from Norwegian wild- and production animals. 

- Identify "main culprit" mechanisms that infer high MIC values for CIP and/or NAL.

- Characterize the population structure of QREC from selected animal species.

## Isolates and Methods

Groups of quinolone resistant *E. coli* (QREC) isolates based on animal species, minimum inhibitory concentration (MIC) values to CIP and NAL, and the amount of additional resistances were created, and isolates were randomly selected from these groups (*n* = 285). This was done to ensure phenotypic diversity in the isolates. Of the 285 isolates, 73 were already sequenced in other projects, and 212 were subjected to whole genome sequencing. MIC values were extracted from the internal sample registration system at the Norwegian Veterinary Institute.

The sequences were subjected to quality control with FastQC. Mash screen was used to identify contaminant reads in the files. Four isolates were excluded from the dataset due to contamination with *Citrobacter* reads. This is probably due to contaminated culture during DNA extraction.

To identify resistance genes of interest, the sequences were run through the program "Antimicrobial Resistance Identification By Assembly" (ARIBA), which is included in the Bifrost Pipeline. Acquired resistance genes were detected by the use of the ResFinder database, while mutations in intrinsic genes related to quinolone resistance was identified with the MegaRes database.
Sequence types was also identified by ARIBA, using the Wirth et al. scheme (*E. coli* #1).

Results from various analyses are listed below.

```{r load libraries, message=FALSE, warning=FALSE}
# Libraries
if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr, ggplot2, tidyr, cowplot, viridis, forcats, radiant.data, purrr, phangorn, gridExtra, ggtree, htmlTable, ape, cluster, tibble)
```

```{r Import data, include = FALSE}
# Collapses data frame to unique lines while ignoring NA's
func_paste <- function(x) paste(unique(x[!is.na(x)]), collapse = ", ")

"%not_in%" <- Negate("%in%")

id_data <- read.table("../data/id.txt",
                      sep = "\t",
                      header = TRUE,
                      stringsAsFactors = F)

isolate_data <- read.table("../data/isolate_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

two_reads <- read.table("../data/no_of_reads.txt",
                          sep = "\t",
                          header = F,
                          stringsAsFactors = F) %>%
  rename("ref" = V1,
         "reads" = V2) %>%
  mutate(ref = gsub(".+/(.*?)\\.fastq\\.gz", "\\1", ref),
         type = ifelse(grepl("R1", ref) == TRUE, "R1", "R2")) %>%
  mutate(ref = gsub("(.*?)_L00[0-9]_R[1-2]_00[0-9]","\\1", ref)) %>%
  group_by(ref) %>%
  mutate(id = n()) %>%
  {. ->> four_reads} %>%
  filter(id == 2) %>%
  spread(type, reads) %>%
  rename("R1-1" = R1,
         "R2-1" = R2) %>%
  mutate("R1-2" = NA,
         "R2-2" = NA) %>%
  select(-id) %>%
  mutate_all(funs(as.character(.)))

four_reads <- four_reads %>%
  filter(id == 4) %>%
  group_by(ref) %>%
  mutate(id = 1:n()) %>%
  spread(id, reads) %>%
  summarise_all(funs(func_paste)) %>%
  rename("R1-1" = `1`,
         "R2-1" = `2`,
         "R1-2" = `3`,
         "R2-2" = `4`) %>%
  select(-type) %>%
  mutate_all(funs(as.character(.)))

no_of_reads <- as.data.frame(rbind(as.matrix(two_reads), as.matrix(four_reads))) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

mic_data <- read.table("../data/mic_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

mlst_data <- read.table("../data/mlst_results.txt", sep = "\t", header = TRUE, stringsAsFactors = F) %>%
  filter(id %not_in% "2014-01-1741-1-S")
```


# Sequence data
## Distribution of number of reads per sequencer
```{r}
sequence_data <- no_of_reads %>%
  left_join(isolate_data, by = "id") %>%
  mutate_at(vars(`R1-1`,`R2-1`,`R1-2`,`R2-2`),
            funs(as.numeric(as.character(.))))

ggplot(sequence_data, aes(as.numeric(`R1-1`)))+
  geom_density(aes(fill = sequencer,
                   y = ..scaled..),
               alpha = 0.6)+
  labs(x = "Number of reads",
       y = "Density",
       fill = "Sequencer")+
  theme(legend.position = c(0.7, 0.8))
```

## Distribution of number of reads per library prep type
```{r}
ggplot(sequence_data, aes(as.numeric(`R1-1`)))+
  geom_density(aes(fill = libprep,
                   y = ..scaled..),
               alpha = 0.6)+
  labs(x = "Number of reads",
       y = "Density",
       fill = "Library prep kit")+
  theme(legend.position = c(0.7, 0.8))
```

## Distribution of number of reads per animal species
```{r}
ggplot(sequence_data, aes(as.numeric(`R1-1`)))+
  geom_density(aes(fill = species,
                   y = ..scaled..),
               alpha = 0.6)+
  labs(x = "Number of reads",
       y = "Density",
       fill = "Animal species")+
  theme(legend.position = c(0.7, 0.8))
```

```{r ARIBA analysis, include = FALSE}
megares_report_loc <- "../data/ariba/megares_results"
resfinder_report_loc <- "../data/ariba/resfinder_results"
ac_genes <- c("qnr","oqx","qep","mfp","mcb")
mut_genes <- c("gyr","par","marR","marA","acrB","soxR","acrR","nor","ydh","sdi","rpoB")

# adjust parameters for filtering
if (length(ac_genes) == 1) {
  if (grepl("all", ac_genes, ignore.case = TRUE) == TRUE) {
    ac_genes <- "ALL"
    } else {
      ac_genes <- ac_genes
    }
}

if (length(mut_genes) == 1) {  
  if (grepl("all", mut_genes, ignore.case = TRUE) == TRUE) {
    mut_genes <- "ALL"
    } else {
      mut_genes <- mut_genes
    }
}
# -------------------------------------- Libraries

if (!require("pacman")) install.packages("pacman")
pacman::p_load(ggplot2,dplyr,tidyr,gridExtra,grid,
               forcats,purrr,stringr,kableExtra,
               knitr,IRdisplay,reprex,svglite)

# -------------------------------------- Functions

func_paste2 <- function(x) paste(unique(x[!is.na(x)]), collapse = ",")

# Confidence interval function
get_binCI <- function(x, n) as.numeric(setNames(binom.test(x,n)$conf.int*100,
                                                c("lwr", "upr")))

# Identifies filenames in input folder
file_names <- function(filepath) {
  files <- list.files(path = filepath, pattern = "amr_report.tsv")
  return(files)
}

# Import ariba data from report.tsv from chosen database used in ariba
get_ariba_data <- function(filepath) {
  files <- file_names(filepath)
  
  data_list <- lapply(files,
                      FUN = function(file) {
                        read.delim(
                          paste0(filepath, "/", file),
                          stringsAsFactors = F,
                          header = TRUE,
                          sep = "\t"
                        )
                      })
  
  names(data_list) <- files
  data <- bind_rows(lapply(data_list, function(x) map(x, as.character)), .id = "ref")
  return(data)
}

# Corrects the gene names found in the "cluster" column
fix_gene_names <- function(df) {
  genes <- unique(df$ref_name)
  new_names <- gsub("^(.*?)\\..*", "\\1", genes)
  new_names <- gsub("_", "", new_names, fixed = T)
  new_names <- gsub("-", "", new_names, fixed = T)
  
  gene_names <- c()
  for (i in new_names) {
    p <- paste(tolower(substring(i, 1,3)), substring(i, 4), sep = "", collapse = " ")
    gene_names <- c(gene_names,p)
  }
  df2 <- data.frame(genes,gene_names) %>%
    mutate(genes = as.character(genes)) %>%
    rename(ref_name = genes)
  
  df <- df %>%
    left_join(df2, by = "ref_name") %>%
    mutate(gene_names = as.character(gene_names),
           ref = gsub("(.*?)_amr_report.tsv", "\\1", ref))
  
  return(df)
}

# Function for selecting genes of interest and filtering the 
# columns in the data frame on the resulting vector
select_genes <- function(df, gene_string) {
  names_df <- names(df)
  x <- "ref"
  for (i in names_df) {
    for (j in gene_string) {
      if (str_detect(i, regex(j, ignore_case = T)) == TRUE) {
        x <- c(x, i)
      }
    }
  }
  df <- df %>%
    select(one_of(x))
  return(df)
}

# Allowed flags from the ARIBA report
flag_selection <- c("19","27","147","155","403",
                    "411","915","923","787","795",
                    "531","539","659","667","787","795")  

# Function that returns info on flag selection
check_flags <- function(df) {
  df <- df %>%
    select(ref, gene_names, flag, ref_ctg_change) %>%
    mutate(flag_result = flag %in% flag_selection,
           flag_result = as.integer(flag_result)) %>%
    rename("gene" = gene_names)
  return(df)
}

# Function that handles megares data and returns a data frame with information on whether a gene
# is mutated or not. Includes control for QRDR in gyrA.
create_mut_table <- function(df) {
  df <- df %>%
    mutate(ref_ctg_change = ifelse(ref_ctg_change == ".", "", ref_ctg_change),
           ref_nt = ifelse(ref_nt == ".", "", ref_nt),
           ctg_nt = ifelse(ctg_nt == ".", "", ctg_nt),
           ref_ctg_nt_change = paste(ref_nt, ctg_nt, sep = "-"),
           mutation = paste(ref_ctg_change, ref_ctg_nt_change, sep = "/"),
           mutation = ifelse(mutation == "/-", "0", mutation)) %>%
    select(ref, gene_names, flag, mutation) %>%
    filter(flag %in% flag_selection) %>%
    mutate(id = 1:n()) %>%
    spread(gene_names, mutation) %>%
    group_by(ref) %>%
    summarise_all(funs(func_paste)) %>%
    select(-c(id, flag)) %>%
    gather(gene, mut, -ref) %>%
    mutate(mut = ifelse(mut == "" | mut == "." | is.na(mut) == TRUE, 0, mut),
           result = ifelse(mut != 0, 1, 0),
           result = as.integer(result),
           type = "mut",
           nt_change = gsub("[^,]+?/([^,]+?)", "\\1", mut),
           mut = gsub("([^/]+)/[^,]+", "\\1", mut))
  return(df)
}                    

# Corrects or mutations in QRDR for gyrA, gyrB, parC and parE genes.
# The function returns columns of 1/0 values for whether or not the 
# mutations reported by ARIBA is within the QRDR in the gene:
# gyrA: AA 67 - 106
# gyrB: AA 333 - 481
# ParC: AA 51 - 170
# parE: AA 366 - 523
# Source: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1142146/pdf/cjvr68pg229.pdf
fix_gyr_par_results <- function(df) {
  df <- df %>%
    mutate(gyrA_result = mut %>% # control for mutation within QRDR for gyrA
             str_extract_all("\\d+") %>% # from reprex package
             map(as.integer) %>% # converts all to integer
             map_lgl(~ any(.x >= 67L & .x <= 106L)), # returns TRUE/FALSE whether value is within range or not
           gyrA_result = if_else(gene != "gyrA", NA, gyrA_result), # filters out results for all other genes
           gyrA_result = as.integer(gyrA_result),
           gyrB_result = mut %>% # control for mutation within QRDR for gyrB
             str_extract_all("\\d+") %>%
             map(as.integer) %>%
             map_lgl(~ any(.x >= 333L & .x <= 481L)),
           gyrB_result = if_else(gene != "gyrB", NA, gyrB_result),
           gyrB_result = as.integer(gyrB_result),
           parC_result = mut %>% # control for mutation within QRDR for parC
             str_extract_all("\\d+") %>% 
             map(as.integer) %>% 
             map_lgl(~ any(.x >= 51L & .x <= 170L)),
           parC_result = if_else(gene != "parC", NA, parC_result),
           parC_result = as.integer(parC_result),
           parE_result = mut %>% # control for mutation within QRDR for parE
             str_extract_all("\\d+") %>% 
             map(as.integer) %>% 
             map_lgl(~ any(.x >= 366L & .x <= 523L)),
           parE_result = if_else(gene != "parE", NA, parE_result),
           parE_result = as.integer(parE_result)) %>%
    mutate(result_gyr_par = case_when(gene == "gyrA" ~ gyrA_result,
                                      gene == "gyrB" ~ gyrB_result,
                                      gene == "parC" ~ parC_result,
                                      gene == "parE" ~ parE_result)) %>%
    mutate(result_total = ifelse(gene %in% c("gyrA","gyrB","parC","parE"), result_gyr_par, result)) %>%
    select(-c(gyrA_result,
              gyrB_result,
              parC_result,
              parE_result,
              result_gyr_par,
              result))
  return(df)
}

# Function that handles resfinder data and returns a data frame with 
# presence/absence for acquired genes                           
create_acquired_table <- function(df) {
  df <- df %>%
    select(ref, gene_names, flag, ref_ctg_change) %>%
    mutate(id = 1:n()) %>%
    filter(flag %in% flag_selection) %>%
    spread(gene_names, ref_ctg_change) %>%
    group_by(ref) %>%
    summarise_all(funs(func_paste)) %>%
    mutate(ref = gsub("^(.*?)_amr_report.tsv$", "\\1", ref)) %>%
    select(-c(id, flag)) %>%
    gather(gene, result,-ref) %>%
    mutate(result_total = ifelse(result == "", 0, 1),
           result_total = as.character(result_total),
           type = "gene") %>%
    select(-result)
  return(df)
}

# Function that returns a filtered dataframe based on the string "genes"
filter_mut_table <- function(df) {
  report_genes <- unique(df$gene)
  grep_genes <- c()
  for (gene in mut_genes) {
    for (g in report_genes) {
      if (grepl(gene, g, ignore.case = T) == TRUE) {
        grep_genes <- c(grep_genes, g)
      }
    }
  }
  df <- df %>%
    filter(gene %in% grep_genes) %>%
    mutate(gene = as.character(gene))
  return(df)
}

# Function that returns a filtered dataframe based on the string "acquired_genes"
filter_acquired_table <- function(df) {
  report_genes <- unique(df$gene)
  grep_genes <- c()
  for (gene in ac_genes) {
    for (g in report_genes) {
      if (grepl(gene, g, ignore.case = T) == TRUE) {
        grep_genes <- c(grep_genes, g)
      }
    }
  }
  df <- df %>%
    filter(gene %in% grep_genes) %>%
    mutate(gene = gsub("_", "", gene),
           gene = as.character(gene))
  return(df)
}

# calculates percentage of present/absent mutations and genes
calc_stats <- function(df) {
  df <- df %>%
    group_by(gene, result_total) %>%
    count() %>%
    ungroup() %>%
    mutate(result_total = if_else(result_total == 1, "Present", "Absent")) %>%
    spread(result_total, n, fill = 0) %>%
    rowwise() %>%
    mutate(Total = Present + Absent,
           Percent = round(Present/Total*100, 1),
           lwr = round(get_binCI(Present, Total)[1], 1),
           upr = round(get_binCI(Present, Total)[2], 1))
  return(df)
}

# calculates how many mutations are present in the genes
# gyrA, gyrB, parC and parE
calc_no_of_mut <- function(df) {
  df1 <- df %>%
    filter(gene %in% c("gyrA","parC","gyrB","parE")) %>%
    mutate(entries = sapply(strsplit(.$mut, ","), FUN = function(x) {length(x)})) %>%
    separate(mut, into = as.character(c(1:max(.$entries)))) %>%
    select(-entries) %>%
    gather(id, mut, -c(ref, gene, result, type))
  
  df2 <- fix_gyr_par_results(df1)
  
  df3 <- df2 %>%
    mutate(test = ifelse(mut != "0" & result_total == 0, 0, 1)) %>%
    filter(test == 1) %>%
    spread(gene, mut) %>%
    group_by(ref) %>%
    summarise_all(funs(func_paste)) %>%
    mutate_at(.vars = vars(c("gyrA","gyrB","parC","parE")),
              .funs = function(x) ifelse(x == "0", "", x)) %>%
    mutate(mut_gyrA = sapply(strsplit(.$gyrA, ","), FUN = function(x) {length(x)}),
           mut_gyrB = sapply(strsplit(.$gyrB, ","), FUN = function(x) {length(x)}),
           mut_parC = sapply(strsplit(.$parC, ","), FUN = function(x) {length(x)}),
           mut_parE = sapply(strsplit(.$parE, ","), FUN = function(x) {length(x)})) %>%
    select(-c(type, id, result_total, test)) %>%
    mutate_at(.vars = vars(c("gyrA","gyrB","parC","parE")),
              .funs = function(x) ifelse(x == "", "0", x))
  return(df3)
}

# creates a data frame with one row per sample, and 1/0 results 
# for mutations in respective genes in columns
create_mut_report <- function(df) {
  df <- df %>%
    select(-mut) %>%
    group_by(id) %>%
    mutate(id2 = 1:n()) %>%
    spread(gene, result_total) %>%
    summarise_all(funs(func_paste)) %>%
    select(-c(id2, type))
  return(df)
}

# creates a data frame with one row per sample, and 1/0 results
# for acquired genes in columns
create_acquired_report <- function(df) {
  df <- df %>%
    group_by(id) %>%
    mutate(id2 = 1:n()) %>%
    spread(gene, result_total) %>%
    summarise_all(funs(func_paste)) %>%
    select(-c(id2, type))
  return(df)
}

# -------------------------------------- Analysis
# Import data
mut_data <- get_ariba_data(megares_report_loc)
acquired_data <- get_ariba_data(resfinder_report_loc)

# Clean data
clean_mut_data <- fix_gene_names(mut_data)
clean_acquired_data <- fix_gene_names(acquired_data)

# Check flags
mut_flags <- check_flags(clean_mut_data) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

acquired_flags <- check_flags(clean_acquired_data) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

# Wrangle
mut_table <- create_mut_table(clean_mut_data)
mut_table_fixed <- fix_gyr_par_results(mut_table) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

mut_filtered <- filter_mut_table(mut_table_fixed)
acquired_table <- create_acquired_table(clean_acquired_data) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

acquired_filtered <- filter_acquired_table(acquired_table)

# Stats
if (length(mut_genes) == 1) {
  if (mut_genes == "ALL") {
    mut_stats <- calc_stats(mut_table_fixed)
    mut_report <- create_mut_report(mut_table_fixed)
    }
} else {
  mut_stats <- calc_stats(mut_filtered)
  mut_report <- create_mut_report(mut_filtered)
}

if (length(ac_genes) == 1) {
  if (ac_genes == "ALL") {
    acquired_stats <- calc_stats(acquired_table)
    acquired_report <- create_acquired_report(acquired_table)
  }
} else {
  acquired_stats <- calc_stats(acquired_filtered)
  acquired_report <- create_acquired_report(acquired_filtered)
}

mut_quant <- suppressWarnings(calc_no_of_mut(mut_table)) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% c("2016-17-565-1-S",
                       "2016-02-428-2-S",
                       "2016-02-486-2-S",
                       "2016-02-732-2-S",
                       "2014-01-1741-1-S"))

# ---------------------------------- Tables

# number of mutations table
mut_no <- mut_quant %>%
  group_by(mut_gyrA, mut_gyrB, mut_parC, mut_parE) %>%
  count()

# number of isolates per mutation combination
mut_comb <- mut_quant %>%
  group_by(gyrA, gyrB, parC, parE) %>%
  count()
```

# Tables
## Isolate data
```{r}
isolate_data_complete <- isolate_data %>%
  left_join(no_of_reads, by = "id")

isolate_data_complete
```

## List of genes of interest
This table lists all the genes of interest in this study, some information about them, and their accession number(s).
```{r}

acquired_ref_names <- acquired_data$ref_name
acquired_ref <- c()

for (i in ac_genes) {
  for (j in acquired_ref_names) {
    if (grepl(i,j, ignore.case = T) == TRUE) {
      acquired_ref <- c(acquired_ref, j)
    }
  }
}

acquired_ref <- unique(acquired_ref)

acquired_ref_genes <- acquired_data %>%
  mutate(test = duplicated(ref_name),
         gene = case_when(grepl("qnra1", ref_name, ignore.case = T) == TRUE ~ "qnrA1",
                          grepl("qnrb19", ref_name, ignore.case = T) == TRUE ~ "qnrB19",
                          grepl("qnrb6", ref_name, ignore.case = T) == TRUE ~ "qnrB6",
                          grepl("qnrb60", ref_name, ignore.case = T) == TRUE ~ "qnrB60",
                          grepl("qnrs1", ref_name, ignore.case = T) == TRUE ~ "qnrS1",
                          grepl("qnrs2", ref_name, ignore.case = T) == TRUE ~ "qnrS2",
                          grepl("qnrs4", ref_name, ignore.case = T) == TRUE ~ "qnrs4",
                          grepl("qnrs9", ref_name, ignore.case = T) == TRUE ~ "qnrS9",
                          grepl("qep", ref_name, ignore.case = T) == TRUE ~ "qepA")) %>%
  filter(test == FALSE,
         ref_name %in% acquired_ref_names) %>%
  select(ref_name, gene) %>%
  mutate(compl = complete.cases(.)) %>%
  filter(compl == TRUE) %>%
  mutate(ref_name2 = sub(".+_(.*?)", "\\1", ref_name)) %>%
  group_by(gene) %>%
  summarise_all(funs(func_paste)) %>%
  select(gene, ref_name2) %>%
  rename("Accession number(s)" = ref_name2,
         "Gene" = gene) %>%
  mutate(Description = "Quinolone resistance protein, plasmid mediated, decreases susceptibility of quinolones",
         Type = "Acquired")
  
df_ref_names <- mut_data$cluster
ref_names <- c()

for (i in mut_genes) {
  for (j in df_ref_names) {
    if (grepl(i,j, ignore.case = T) == TRUE) {
      ref_names <- c(ref_names, j)
    }
  }
}

ref_names <- unique(ref_names)

intrinsic_ref_genes <- mut_data %>%
  mutate(test = duplicated(cluster),
         gene = case_when(grepl("gyra", cluster, ignore.case = T) == TRUE ~ "gyrA",
                          grepl("gyrb", cluster, ignore.case = T) == TRUE ~ "gyrB",
                          grepl("parc", cluster, ignore.case = T) == TRUE ~ "parC",
                          grepl("pare", cluster, ignore.case = T) == TRUE ~ "parE",
                          grepl("soxR", cluster, ignore.case = T) == TRUE ~ "soxR",
                          grepl("marR", cluster, ignore.case = T) == TRUE ~ "marR",
                          grepl("mara", cluster, ignore.case = T) == TRUE ~ "marA",
                          grepl("tolc", cluster, ignore.case = T) == TRUE ~ "tolC",
                          grepl("acrr", ref_name, ignore.case = T) == TRUE ~ "acrR",
                          grepl("acrb", ref_name, ignore.case = T) == TRUE ~ "acrB",
                          grepl("mdfa", ref_name, ignore.case = T) == TRUE ~ "mdfA",
                          grepl("rpob", ref_name, ignore.case = T) == TRUE ~ "rpoB")) %>%
  filter(test == FALSE,
         cluster %in% ref_names) %>%
  select(ref_name, gene, free_text) %>%
  mutate(ref_name2 = sub("^.*?_([A-Z]+[0-9_.]*[0-9]).*", "\\1", ref_name)) %>%
  select(-ref_name) %>%
  group_by(gene) %>%
  summarise_all(funs(func_paste)) %>%
  select(gene, ref_name2) %>%
  rename("Accession number(s)" = ref_name2,
         "Gene" = gene) %>%
  mutate(Description = case_when(Gene == "acrR" ~ "Multidrug efflux regulator, represses the transcription of acrB.",
                                 Gene == "acrB" ~ "Gene that codes for the AcrB efflux pump, part of the AcrAB-TolC efflux system",
                                 Gene == "gyrA" ~ "DNA gyrase subunit A, regulates DNA topoplogy. Target for quinolones.",
                                 Gene == "gyrB" ~ "DNA gyrase subunit B, regulates DNA topoplogy. Target for quinolones.",
                                 Gene == "marR" ~ "Repressor of the marRAB operon, involved with the activation of resistance- and oxidative stress genes.",
                                 Gene == "marA" ~ "Transcriptional activator og genes involved in the MAR phenotype.",
                                 Gene == "parC" ~ "Topoisomerase IV subunit C, regulates DNA topology. Target for quinolones.",
                                 Gene == "parE" ~ "Topoisomerase IV subunit E, regulates DNA topology. Target for quinolones.",
                                 Gene == "soxR" ~ "Transcriptional activator of the superoxide response regulon.",
                                 Gene == "tolC" ~ "Part of the major AcrAB-TolC multidrug efflux pump.",
                                 Gene == "mdfA" ~ "Chromosomal efflux pump.",
                                 Gene == "rpoB" ~ "Beta-subunit of RNA polymerase, mutations lead to increased expression of the ydhE-pump"),
         Type = "Intrinsic gene")
                
ref_genes <- rbind(intrinsic_ref_genes, acquired_ref_genes)

ref_genes

```

## Isolates per species
```{r}
per_species <- isolate_data %>%
  group_by(species) %>%
  count()

per_species
```

## Percent isolates with PMQR
```{r}
acquired_report %>%
  select(-id) %>%
  mutate_all(funs(as.numeric(.))) %>%
  mutate(total = rowSums(.),
         result = if_else(total != 0, 1, 0)) %>%
  group_by(result) %>%
  count() %>%
  ungroup() %>%
  mutate(result = if_else(result == 0, "Absent", "Present")) %>%
  spread(result, n) %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total * 100,1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr))
```

## Percent isolates with specific PMQR genes
```{r}
acquired_report %>%
  select(-id) %>%
  t() %>%
  as.data.frame(.) %>%
  mutate(Gene = rownames(.)) %>%
  mutate_at(vars(-Gene),
            funs(as.numeric(as.character(.)))) %>%
  mutate(Present = rowSums(.[,-281])) %>%
  select(c(Gene, Present)) %>%
  rowwise() %>%
  mutate(Total = 280,
         Percent = round(Present/Total * 100,1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr))
```

## Percent isolates with mutations in QRDR
```{r}
mut_report %>%
  select(gyrA, gyrB, parC, parE) %>%
  group_by_all() %>%
  count() %>%
  ungroup() %>%
  mutate_at(vars(c(gyrA, gyrB, parC, parE)),
            .funs = funs(as.numeric)) %>%
  mutate(type = ifelse(rowSums(.[,1:4]) == 0, 1, 0)) %>%
  select(n, type) %>%
  group_by(type) %>%
  summarise_all(funs(sum)) %>%
  mutate(type = ifelse(type == 0, "Present", "Absent")) %>%
  spread(type, n) %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total * 100,1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr, Total))
```

## Percent of isolates with specific mutation combinations in QRDR
This table presents the different substitution combinations identified in the quinolone resistance determining regions (QRDR) of *gyrA*, *gyrB*, *parC* and *parE*.
```{r}
mut_quant %>%
  group_by(gyrA, gyrB, parC, parE) %>%
  count() %>%
  rowwise() %>%
  mutate(Total = 280,
         Percent = round(n/Total * 100,1),
         lwr = round(get_binCI(n, Total)[1], 1),
         upr = round(get_binCI(n, Total)[2], 1),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr, Total))
```

## Percent isolates with specific nucleotide changes in QRDR
This table presents the various nucleotide changes for each amino acid substitution identified in the quinolone resistance determining region (QRDR) of *gyrA*, *gyrB*, *parC* and *parE*.
```{r}
suppressWarnings(clean_mut_data %>%
  mutate(ref_ctg_change = ifelse(ref_ctg_change == ".", "", ref_ctg_change),
         ref_nt = ifelse(ref_nt == ".", "", ref_nt),
         ctg_nt = ifelse(ctg_nt == ".", "", ctg_nt),
         ref_ctg_nt_change = paste(ref_nt, ctg_nt, sep = "-"),
         mutation = paste(ref_ctg_change, ref_ctg_nt_change, sep = "/"),
         mutation = ifelse(mutation == "/-", "0", mutation)) %>%
  select(ref, gene_names, flag, mutation) %>%
  filter(flag %in% flag_selection,
         gene_names %in% c("gyrA","gyrB","parC","parE")) %>%
  mutate(id = 1:n()) %>%
  spread(gene_names, mutation) %>%
  group_by(ref) %>%
  summarise_all(funs(func_paste2)) %>%
  select(-c(id, flag)) %>%
  gather(gene, mut, -ref) %>%
  mutate(mut = ifelse(mut == "" | mut == "." | is.na(mut) == TRUE, 0, mut),
         result = ifelse(mut != 0, 1, 0),
         result = as.integer(result),
         type = "mut") %>%
  separate(mut, into = paste("v", 1:12, sep = "_"), sep = ",") %>%
  gather(mut, value, paste("v", 1:12, sep = "_")) %>%
  separate(value, into = c("aa_change","nt_change"), sep = "/") %>%
  select(-mut) %>%
  rename("mut" = aa_change) %>%
  filter(is.na(nt_change) == FALSE) %>%
  fix_gyr_par_results(.) %>%
  filter(result_total == 1) %>%
  group_by(gene, mut, nt_change) %>%
  count() %>%
  separate(nt_change, into = c("Reference nt", "Contig nt"), sep = "-") %>%
  rowwise() %>%
  mutate(Total = 280,
         Percent = round(n/Total * 100,1),
         lwr = round(get_binCI(n, Total)[1],1),
         upr = round(get_binCI(n, Total)[2],1),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr, Total)) %>%
  rename("Mutation" = mut))
```

## Total combinations
This table presents the total combination of resistance mechanisms, their respective MIC-range for CIP and NAL, how many isolates per group, as well as percentage and confidence intervals.
```{r}
acquired_report %>%
  left_join(mut_quant[c("gyrA","gyrB","parC","parE","id")], by = "id") %>%
  left_join(mut_report[c("id","marR","marA","acrB","rpoB")], by = "id") %>%
  left_join(mic_data[c("id","CIP","NAL")], by = "id") %>%
  left_join(isolate_data[c("id","species")], by = "id") %>%
  select(-id) %>%
  group_by_at(vars(-c(CIP, NAL))) %>%
  summarise(CIP = ifelse(min(as.numeric(CIP)) == max(as.numeric(CIP)),
                         as.character(median(as.numeric(CIP))),
                         paste(min(as.numeric(CIP)), max(as.numeric(CIP)), sep = " - ")),
            NAL = ifelse(min(as.numeric(NAL)) == max(as.numeric(NAL)),
                         as.character(median(as.numeric(NAL))),
                         paste(min(as.numeric(NAL)), max(as.numeric(NAL)), sep = " - ")),
            n = n()) %>%
  rowwise() %>%
  mutate(Total = 280,
         Percent = round(n/Total * 100,1),
         lwr = round(get_binCI(n, Total)[1],1),
         upr = round(get_binCI(n, Total)[2],1),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr, Total))
```

## Amount of isolates resistant to specific antimicrobials
```{r}
mic_data %>%
  mutate(SMX_R = ifelse(SMX > 64, 1, 0),
         TMP_R = ifelse(TMP > 2, 1, 0),
         CIP_R = ifelse(CIP > 0.06, 1, 0),
         TET_R = ifelse(TET > 8, 1, 0),
         NAL_R = ifelse(NAL > 16, 1, 0),
         CTX_R = ifelse(CTX > 0.25, 1, 0),
         CHL_R = ifelse(CHL > 16, 1, 0),
         AMP_R = ifelse(AMP > 8, 1, 0),
         GEN_R = ifelse(GEN > 2, 1, 0)) %>%
  select(contains("_R")) %>%
  gather(AB, value) %>%
  group_by(AB) %>%
  summarise_all(funs(sum)) %>%
  arrange(desc(value)) %>%
  mutate(AB = sub("_R", "", AB))
```

## Sequence types
```{r}
mlst_data[,c("id","ST")] %>%
  group_by(ST) %>%
  count()
```

## Sequence types per animal species
```{r}
mlst_data[,c("id","ST")] %>%
  left_join(isolate_data[,c("id","species")], by = "id") %>%
  group_by(ST, species) %>%
  count()
```

## Significant differences between resistance mechanisms for each animal species
```{r, warning = F, message = F}
AddCol <- function(df, col_name) {
  # split rows by delimeters
  string_to_proc <- df %>% select(!!col_name) %>%
    unlist() %>% str_split(regex("\\, |\\,")) 
  # find unique entries
  unique_strings <- string_to_proc %>%
    unlist() %>% unique()
  # construct names of the new columns
  cols_names <- paste(col_name, unique_strings, sep = "_")
  # construct 0/1-content columns for each unique entry
  cols_content <- sapply(function(i) {
    as.integer(unlist(lapply(function(Z) any(Z %in% unique_strings[i]), 
                             X = string_to_proc)))
  }, X = seq_along(unique_strings))
  res <- data.frame(cols_content)
  names(res) <- cols_names
  return(res)
}

do_chisq_test <- function(df) {
  df <- df %>%
    slice(expand.grid(1:length(unique(species)), 1:length(unique(species))) %>% rev %>% 
          filter(Var2 < Var1) %>% t) %>%
    mutate(test = rep(1:(n()/2), each = 2)) %>%
    group_by(test) %>%
    do(data_frame(gene = .$key,
                  test = first(.$test),
                  species1 = first(.$species),
                  species2 = last(.$species),
                  data = list(matrix(c(.$Absent,
                                       .$Present),
                                     ncol = 2)))) %>%
    mutate(chi_test = map(data, chisq.test, correct = FALSE)) %>%
    mutate(p.value = map_dbl(chi_test, "p.value"),
           dupl = duplicated(test)) %>%
    filter(dupl == FALSE) %>%
    select(gene, species1, species2, p.value) %>%
    mutate(p.value = round(p.value, 5)) %>%
    filter(p.value <= 0.05)
  return(df)
}

total_df <- acquired_report %>%
  left_join(mut_quant[c("gyrA","gyrB","parC","parE","id")], by = "id") %>%
  left_join(mut_report[c("id","marR","marA","acrB","rpoB")], by = "id") %>%
  left_join(isolate_data[c("id", "species")])

col_proc <- c("gyrA","gyrB","parC","parE")

id <- total_df$id

cols_to_add <- sapply(function(i) {AddCol(df = total_df, col_name = col_proc[i])},
                      X = seq_along(col_proc)) %>%
  bind_cols() %>%
  cbind(id) %>%
  gather(key, value, -id) %>%
  filter(grepl("(.*?)_0", key) == FALSE) %>%
  spread(key, value)

mechanism_per_species <- total_df %>%
  select(-c(gyrA, gyrB, parC, parE)) %>%
  left_join(cols_to_add, by = "id") %>%
  select(id, species, everything()) %>%
  select(-id) %>%
  gather(key, value, -species) %>%
  group_by_all() %>%
  count() %>%
  ungroup() %>%
  mutate(value = if_else(value == 1, "Present", "Absent")) %>%
  spread(value, n, fill = 0) %>%
  select(key, species, Present, Absent) %>%
  split(f = .$key)

p_values_total <- suppressWarnings(lapply(mechanism_per_species, function(x) do_chisq_test(x))) %>%
  bind_rows()

p_values_total
```

# Figures
## Percent isolates with PMQR and/or QRDR mutations in *gyrA*, *gyrB*, *parC* or *parE*
```{r,message = FALSE, warning = FALSE}
total_report <- mut_report %>%
  left_join(acquired_report, by = "id") %>%
  gather(gene, value, -id) %>%
  mutate(present = ifelse(value == 1, gene, NA),
         ref = 1:n()) %>%
  spread(gene, value) %>%
  group_by(id) %>%
  summarise_all(funs(func_paste)) %>%
  select(-ref) %>%
  left_join(mic_data[c("id","CIP")], by = "id") %>%
  group_by(present,CIP) %>%
  count() %>%
  ungroup() %>%
  rowwise() %>%
  mutate(present = ifelse(present == "", "None", present),
         total = 280,
         percent = round(n/total * 100, 1),
         lwr = round(get_binCI(n, total)[1], 1),
         upr = round(get_binCI(n, total)[2], 1))

genes <- c(names(mut_report), names(acquired_report))
genes <- genes[!genes %in% c("id","nt_change")]

cip_n_plot <- ggplot(total_report, aes(reorder(present, as.numeric(CIP)), percent, fill = factor(CIP)))+
  geom_col(color = "black")+
  labs(y = "Percent (%) isolates")+
  scale_fill_viridis(discrete = TRUE)+
  guides(fill=guide_legend(ncol = 10))+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 8),
        axis.title.x = element_blank(),
        legend.text = element_text(size = 8),
        legend.title = element_blank(),
        legend.position = "top",
        legend.spacing.x = unit(0.2, "cm"),
        legend.justification = "center")

dot_report <- mut_report %>%
  left_join(acquired_report, by = "id") %>%
  select(-nt_change) %>%
  gather(gene, value, -id) %>%
  mutate(present = ifelse(value == 1, gene, NA),
         ref = 1:n()) %>%
  spread(gene, value) %>%
  group_by(id) %>%
  summarise_all(funs(func_paste)) %>%
  left_join(mic_data[c("id","CIP")], by = "id") %>%
  select(-c(ref, id)) %>%
  ungroup() %>%
  mutate_at(vars(-present, CIP), .funs = as.numeric) %>%
  mutate(none = ifelse(rowSums(.[,genes]) == 0, 1, 0),
         present = ifelse(present == "", "None", present)) %>%
  gather(gene, value, -c(present,CIP)) %>%
  mutate(gene = ifelse(value == 0 & gene != "None", NA, gene),
         value = ifelse(gene == "None" & value == 0, NA, value)) %>%
  na.omit()

  
mechanism_dot_plot <- ggplot(dot_report, aes(reorder(present, CIP), gene, fill = gene))+
  geom_point(size = 2, pch = 21)+
  scale_fill_viridis(option = "B",
                     discrete = TRUE)+
  guides(fill = FALSE)+
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.title = element_blank(),
        axis.ticks.x = element_blank())

plot_grid(cip_n_plot, mechanism_dot_plot, nrow = 2, align = "v", rel_heights = c(3.5/5, 1.5/5))

```

## Percent QREC isolates with mutations in intrinsic genes
```{r}
gene_names <- names(mut_report)

gene_names <- gene_names[-c(1,2)]

palette <- c("Broiler" = "#4575b4",
             "Pig" = "#74add1",
             "Red fox" = "#f46d43",
             "Wild bird" = "#fdae61")

mut_report %>%
    left_join(isolate_data, by = "id") %>%
    gather(key, value, gene_names) %>%
    group_by(key, value, species) %>%
    count() %>%
    ungroup() %>%
    mutate(value = if_else(value == 1, "Present", "Absent")) %>%
    spread(value, n, fill = 0) %>%
    rowwise() %>%
    mutate(
      Total = Present + Absent,
      Percent = round(Present / Total * 100, 1),
      lwr = round(get_binCI(Present, Total)[1], 1),
      upr = round(get_binCI(Present, Total)[2], 1)
    ) %>%
    mutate(species = factor(species,
                            levels = c("Broiler","Pig","Wild bird","Red fox"))) %>%
    ggplot(aes(key, Percent, fill = species)) +
    geom_col(color = "black",
             position = position_dodge(1)) +
    geom_errorbar(
      aes(ymin = lwr, ymax = upr),
      width = 0.6,
      position = position_dodge(1)
    ) +
    scale_fill_manual(values = palette) +
    scale_x_discrete(limits = c("gyrA","gyrB","parC","parE","marA","marR","acrB","rpoB")) +
    labs(y = "Percent (%) of isolates") +
    guides(fill = FALSE) +
    theme_classic() +
    theme(
      axis.text.x = element_text(
        angle = 90,
        size = 12,
        hjust = 1,
        vjust = 0.4
      ),
      axis.title.x = element_blank(),
      axis.text.y = element_text(size = 12),
      axis.title.y = element_text(size = 14),
      strip.text = element_text(size = 12)
    ) +
    facet_wrap(~ species)
```

## Percent QREC isolates with Qnr genes per species
```{r}
qnr_genes <- c("qnrA1","qnrB19","qnrB6","qnrB60","qnrS1","qnrS2","qnrS4")

acquired_report %>%
    left_join(isolate_data, by = "id") %>%
    gather(key, value, qnr_genes) %>%
    group_by(key, value, species) %>%
    count() %>%
    ungroup() %>%
    mutate(value = if_else(value == 1, "Present", "Absent")) %>%
    spread(value, n, fill = 0) %>%
    rowwise() %>%
    mutate(
      Total = Present + Absent,
      Percent = round(Present / Total * 100, 1),
      lwr = round(get_binCI(Present, Total)[1], 1),
      upr = round(get_binCI(Present, Total)[2], 1)
    ) %>%
    ggplot(aes(key, Percent, fill = species)) +
    geom_col(color = "black",
             position = position_dodge(0.9)) +
    geom_errorbar(
      aes(ymin = lwr, ymax = upr),
      width = 0.6,
      position = position_dodge(0.9)
    ) +
    scale_fill_manual(values = palette) +
    labs(y = "Percent (%) of isolates",
         fill = NULL) +
    theme_classic() +
    theme(
      axis.text = element_text(size = 12),
      axis.title.y = element_text(size = 14),
      legend.text = element_text(size = 12),
      axis.title.x = element_blank(),
      legend.justification = c(0, 1),
      legend.position = c(0.82, 0.97)
    )
```

## Total resistance mechanisms per animal species
This figure presents the percent combination of resistance mechanisms per CIP MIC category. The X-axis on the leftmost plot lists all the specific mutations in *gyrA*, *parC* and *parE*, whether *marR* is mutated or not, as well as a summation of all the *qnr* genes and *qepA*. The rightmost plot presents how many isolates are present in the given CIP-MIC group (Y-axis in both plots).
```{r, warning=F, message=F, fig.height=16, fig.width=20}
AddCol <- function(df, col_name) {
  # split rows by delimeters
  string_to_proc <- df %>% select(!!col_name) %>%
    unlist() %>% str_split(regex("\\, |\\,")) 
  # find unique entries
  unique_strings <- string_to_proc %>%
    unlist() %>% unique()
  # construct names of the new columns
  cols_names <- paste(col_name, unique_strings, sep = "_")
  # construct 0/1-content columns for each unique entry
  cols_content <- sapply(function(i) {
    as.integer(unlist(lapply(function(Z) any(Z %in% unique_strings[i]), 
                             X = string_to_proc)))
  }, X = seq_along(unique_strings))
  res <- data.frame(cols_content)
  names(res) <- cols_names
  return(res)
}

total_df <- acquired_report %>%
  left_join(mut_quant[c("gyrA","gyrB","parC","parE","id")], by = "id") %>%
  left_join(mut_report[c("id","marR","marA","acrB","rpoB")], by = "id") %>%
  left_join(isolate_data[c("id", "species")])

col_proc <- c("gyrA","gyrB","parC","parE")

id <- total_df$id

cols_to_add <- sapply(function(i) {AddCol(df = total_df, col_name = col_proc[i])},
                      X = seq_along(col_proc)) %>%
  bind_cols() %>%
  cbind(id) %>%
  gather(key, value, -id) %>%
  filter(grepl("(.*?)_0", key) == FALSE) %>%
  spread(key, value)

total_df2 <- total_df %>%
  select(-c(gyrA, gyrB, parC, parE)) %>%
  left_join(cols_to_add, by = "id") %>%
  select(id, species, everything()) %>%
  mutate_at(vars(qnrA1,qnrB19,qnrB6,qnrB60,qnrS1,qnrS2,qnrS4),
            funs(as.numeric(.))) %>%
  mutate(qnr = ifelse(rowSums(.[,4:10]) == 0, 0, 1)) %>%
  select(-c(qnrA1,qnrB19,qnrB6,qnrB60,qnrS1,qnrS2,qnrS4)) %>%
  left_join(mic_data[c("id","CIP")], by = "id") %>%
  select(-id) %>%
  mutate_at(vars(-species, CIP),
            funs(as.numeric(as.character(.)))) %>%
  group_by(species, CIP) %>%
  summarise_all(funs(sum, n = n())) %>%
  select(-c(qepA4_n,qnr_n,gyrA_D87G_n,gyrA_D87H_n,
            gyrA_D87N_n,gyrA_D87Y_n,gyrA_S83A_n,
            gyrA_S83L_n,parC_A56T_n,parC_E84V_n,
            parC_S57T_n,parC_S58I_n,parC_S80I_n,
            parC_S80R_n,parE_A512T_n,parE_D463N_n,
            parE_D475E_n,parE_H516Y_n,parE_L416F_n,
            parE_L488M_n,parE_S458A_n,marA_n,rpoB_n,
            acrB_n)) %>%
  gather(gene, value, -c(species, CIP, marR_n)) %>%
  rename("n" = marR_n) %>%
  mutate(gene2 = sub("(.*?)_(.*?)_sum", "\\1", gene),
         gene2 = sub("(.*?)_sum", "\\1", gene2),
         gene = sub("(.*?)_(.*?)_sum", "\\2", gene),
         gene = sub("(.*?)_sum", "\\1", gene),
         type = case_when(gene2 == "gyrA" ~ 1,
                          gene2 == "gyrB" ~ 2,
                          gene2 == "parC" ~ 3,
                          gene2 == "parE" ~ 4,
                          gene2 == "marR" ~ 5,
                          gene2 == "marA" ~ 6,
                          gene2 == "mdfA" ~ 7,
                          gene2 == "acrB" ~ 8,
                          gene2 == "rpoB" ~ 9,
                          gene2 == "qnr" ~ 10,
                          gene2 == "qepA4" ~ 11),
         perc = value/n*100)

p1 <- ggplot(total_df2, aes(reorder(gene, type), factor(CIP), fill = perc))+
  geom_tile(color = "grey70")+
  scale_fill_viridis(option = "A",
                     begin = 0.1,
                     guide = guide_colorbar(barheight = 15,
                                            barwidth = 1,
                                            label.position = "left",
                                            label.hjust = 0.5,
                                            label.vjust = 0.3,
                                            ticks = TRUE))+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.3, size = 18),
        axis.text.y = element_text(size = 18),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size = 20),
        strip.background = element_rect(colour = "black", fill = "grey70"),
        strip.text = element_text(size = 18),
        legend.position = "left",
        legend.title = element_blank(),
        legend.text = element_text(size = 14),
        plot.margin = unit(c(0,0,0,0), "cm"))+
  labs(y = "CIP MIC")+
  facet_wrap(~species, ncol = 1)+
  coord_fixed()

nplot_data <- total_df %>%
  left_join(mic_data[c("id","CIP")], by = "id") %>%
  group_by(species, CIP) %>%
  count() %>%
  ungroup() %>%
  mutate(species = paste0(species, " n"))

p2 <- ggplot(nplot_data, aes(factor(CIP), n))+
  geom_col()+
  theme(axis.title = element_blank(),
        axis.text.y = element_blank(),
        axis.text.x = element_text(size = 18),
        axis.ticks.y = element_blank(),
        strip.text = element_text(size = 18),
        strip.background = element_rect(colour = "black", fill = "grey70"),
        plot.margin = unit(c(0,0,-10,0), "cm"))+
  facet_wrap(~species, ncol = 1)+
  coord_flip()

p1g <- ggplotGrob(p1)
p2g <- ggplotGrob(p2)

g <- cbind(p1g, p2g, size = "first")

grid.arrange(g)

```

## Correlation plot
This plot presents the results of a pearson correlation test for each resistance mechanism. A positive correlation is represented in yellow and red spectra, a negative correlation represented as blue and light blue, and no correlation in green.
```{r,message = FALSE, warning = FALSE}
create_corr_matrix <- function(df) {
  df <- df %>%
    as.matrix %>%
    cor %>%
    as.data.frame %>%
    rownames_to_column(var = "var1") %>%
    gather(var2, value, -var1) %>%
    mutate(var1 = gsub("(.*?)_R$", "\\1", var1),
           var2 = gsub("(.*?)_R$", "\\1", var2))
  return(df)
}

total_df <- acquired_report %>%
  left_join(mut_quant[c("gyrA","gyrB","parC","parE","id")], by = "id") %>%
  left_join(mut_report[c("id","marR","marA","acrB","rpoB")], by = "id")

col_proc <- c("gyrA","gyrB","parC","parE")

id <- total_df$id

cols_to_add <- sapply(function(i) {AddCol(df = total_df, col_name = col_proc[i])},
                      X = seq_along(col_proc)) %>%
  bind_cols() %>%
  cbind(id) %>%
  gather(key, value, -id) %>%
  filter(grepl("(.*?)_0", key) == FALSE) %>%
  spread(key, value)

corr_df <- total_df %>%
  select(-c(gyrA, gyrB, parC, parE)) %>%
  left_join(cols_to_add, by = "id") %>%
  select(-id) %>%
  mutate_all(funs(as.numeric)) %>%
  mutate(qnr = ifelse(rowSums(.[,2:8]) == 0, 0, 1)) %>%
  select(-c(qnrA1,qnrB19,qnrB6,qnrB60,qnrS1,qnrS2,qnrS4))

corr_matrix <- create_corr_matrix(corr_df) %>%
    mutate(gene1 = gsub("(.*?)_.+", "\\1", var1),
           gene2 = gsub("(.*?)_.+", "\\1", var2),
           var1 = gsub(".+_(.*?)", "\\1", var1),
           var2 = gsub(".+_(.*?)", "\\1", var2),
           type1 = case_when(gene1 == "gyrA" ~ 1,
                          gene1 == "gyrB" ~ 2,
                          gene1 == "parC" ~ 3,
                          gene1 == "parE" ~ 4,
                          gene1 == "marR" ~ 5,
                          gene1 == "marA" ~ 6,
                          gene1 == "acrB" ~ 7,
                          gene1 == "rpoB" ~ 8,
                          gene1 == "mdfA" ~ 9,
                          gene1 == "qnr" ~ 10,
                          gene1 == "qepA4" ~ 11),
           type2 = case_when(gene2 == "gyrA" ~ 1,
                          gene2 == "gyrB" ~ 2,
                          gene2 == "parC" ~ 3,
                          gene2 == "parE" ~ 4,
                          gene2 == "marR" ~ 5,
                          gene2 == "marA" ~ 6,
                          gene2 == "acrB" ~ 7,
                          gene2 == "rpoB" ~ 8,
                          gene2 == "mdfA" ~ 9,
                          gene2 == "qnr" ~ 10,
                          gene2 == "qepA4" ~ 11))

cols <- c("#e7f0fa","#c9e2f6","#95cbee","#0099dc","#4ab04a",
          "#ffd73e","#eec73a","#e29421","#f05336","#ce472e")

ggplot(corr_matrix, aes(reorder(var1, type1), reorder(var2, type2), fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradientn(colors=cols,
                       na.value="grey95",
                       limits=c(-1, 1),
                       guide=guide_colourbar(ticks=T,nbin=10,
                                             barheight=.5,label=T, 
                                             barwidth=10))+
  labs(fill = NULL)+
  theme_classic()+
  theme(axis.text = element_text(size = 8),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4),
        axis.title = element_blank(),
        legend.position="bottom",
        legend.justification="center",
        legend.direction="horizontal")+
  coord_fixed()
```

## Sequence types per resistance category
This plot presents the percent of isolates in each sequence type group that are resistant to multiple antimicrobials, sorted from least resistant at the top and most resistant at the bottom. The percentage is based on each ST group, and how many isolates in that group.
```{r}
ST_MIC <- mic_data %>%
  left_join(mlst_data[c("id","ST")], by = "id") %>%
  mutate(SMX_R = ifelse(as.numeric(SMX) > 64, 1, 0),
         TMP_R = ifelse(as.numeric(TMP) > 2, 1, 0),
         CIP_R = ifelse(as.numeric(CIP) > 0.06, 1, 0),
         TET_R = ifelse(as.numeric(TET) > 8, 1, 0),
         NAL_R = ifelse(as.numeric(NAL) > 16, 1, 0),
         CTX_R = ifelse(as.numeric(CTX) > 0.25, 1, 0),
         CHL_R = ifelse(as.numeric(CHL) > 16, 1, 0),
         AMP_R = ifelse(as.numeric(AMP) > 8, 1, 0),
         GEN_R = ifelse(as.numeric(GEN) > 2, 1, 0)) %>%
  mutate(NAL_CIP_R = rowSums(.[,c(15,17)]),
         antres = rowSums(.[,13:21])-NAL_CIP_R) %>%
  select(-c(NAL_CIP_R,SMX,TMP,CIP,TET,NAL,CTX,CHL,AMP,GEN)) %>%
  left_join(isolate_data[c("id", "species")], by = "id") %>%
  gather(AB, res, -c(id, antres, ST, species)) %>%
  group_by(AB, ST, res, species) %>%
  count() %>%
  spread(res, n, fill = 0) %>%
  rename("res" = "1",
         "nonres" = "0") %>%
  group_by(AB, ST) %>%
  mutate(p = res/(res+nonres)*100) %>%
  ungroup() %>%
  mutate(AB = gsub("_R", "", AB))


ggplot(ST_MIC, aes(reorder(AB, -p), reorder(ST, -p), fill = p))+
  geom_tile()+
  scale_fill_viridis(guide = guide_colorbar(barheight = 5,
                                            barwidth = 0.5,
                                            label.position = "left",
                                            label.hjust = 0.5,
                                            label.vjust = 0.3,
                                            ticks = TRUE))+
  labs(y = "Antimicrobials")+
  theme(axis.text.y = element_text(size = 6),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4, size = 8),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        legend.position = "left",
        legend.title = element_blank(),
        legend.text = element_text(size = 7))+
  facet_wrap(~ species, scales = "free_y", ncol = 2)
```

## Phenotypic resistance combinations
This plot presents the various phenotypic resistance combinations present in the isolates, and how many isolates for each animal species in each group.
```{r, fig.width=12, fig.height=10}
mic_dotplot_data <- mic_data %>%
  mutate(SMX_R = ifelse(SMX > 64, 1, 0),
         TMP_R = ifelse(TMP > 2, 1, 0),
         CIP_R = ifelse(CIP > 0.06, 1, 0),
         TET_R = ifelse(TET > 8, 1, 0),
         NAL_R = ifelse(NAL > 16, 1, 0),
         CTX_R = ifelse(CTX > 0.25, 1, 0),
         CHL_R = ifelse(CHL > 16, 1, 0),
         AMP_R = ifelse(AMP > 8, 1, 0),
         GEN_R = ifelse(GEN > 2, 1, 0)) %>%
  select(contains("_R"), -c(CIP_R, NAL_R)) %>%
  mutate(None = ifelse(rowSums(.) == 0, 1, 0)) %>%
  group_by_all() %>%
  count() %>%
  ungroup() %>%
  mutate(id = 1:n()) %>%
  gather(AM, value, -id, -n) %>%
  mutate(value = ifelse(AM == "None" & value == 0, NA, value),
         value = ifelse(AM != "None" & value == 0, NA, value),
         AM = sub("_R", "", AM)) %>%
  filter(is.na(value) == FALSE) %>%
  mutate(AM = factor(AM, ordered = TRUE, levels = c("SMX","TMP","AMP","TET","CHL","CTX","GEN","None")))


mic_nplot_data <- mic_data %>%
  mutate(SMX_R = ifelse(SMX > 64, 1, 0),
         TMP_R = ifelse(TMP > 2, 1, 0),
         CIP_R = ifelse(CIP > 0.06, 1, 0),
         TET_R = ifelse(TET > 8, 1, 0),
         NAL_R = ifelse(NAL > 16, 1, 0),
         CTX_R = ifelse(CTX > 0.25, 1, 0),
         CHL_R = ifelse(CHL > 16, 1, 0),
         AMP_R = ifelse(AMP > 8, 1, 0),
         GEN_R = ifelse(GEN > 2, 1, 0)) %>%
  select(id, contains("_R"), -c(CIP_R, NAL_R)) %>%
  mutate(None = ifelse(rowSums(.[,2:8]) == 0, 1, 0)) %>%
  left_join(isolate_data[c("id", "species")], by = "id") %>%
  select(-id) %>%
  group_by_all() %>%
  count() %>%
  group_by_at(vars(SMX_R, TMP_R, TET_R, CTX_R, CHL_R, AMP_R, GEN_R)) %>%
  { mutate(ungroup(.), group = group_indices(.)) } %>%
  group_by(group) %>%
  mutate(group_n = sum(n))


mic_dotplot <- ggplot(mic_dotplot_data, aes(reorder(factor(id), -n), AM, fill = AM)) +
  geom_point(size = 6, pch = 21)+
  scale_fill_viridis(option = "A",
                     discrete = TRUE)+
  guides(fill = FALSE)+
  theme(axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid.major.x = element_line())

mic_nplot <- ggplot(mic_nplot_data, aes(reorder(factor(group), -group_n), n, fill = species))+
  geom_col(color = "black")+
  scale_fill_manual(values = palette) +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.position = c(0.85,0.85),
        legend.title = element_blank())

cowplot::plot_grid(mic_nplot, mic_dotplot, ncol = 1, align = "v")

```

# Phylogeny
## Core Genome MLST
```{r, fig.width=16, fig.height=14, message = FALSE, warning = FALSE}
cgMLSTdata <- read.table("../data/cgMLST.txt", sep = "\t", row.names=1, colClasses = "factor", header = T)

palette <- c("Broiler" = "#4575b4",
             "Pig" = "#74add1",
             "Red fox" = "#f46d43",
             "Wild bird" = "#fdae61")

cgMLST_tree <- as.phylo(hclust(daisy(cgMLSTdata, metric="gower")))

ids <- as.data.frame(cgMLST_tree$tip.label)
names(ids) <- "id"

ids2 <- ids %>%
  left_join(isolate_data[c("id","species")]) %>%
  left_join(mlst_data[c("id", "ST")])

cgMLST_tree$tip.label <- ids2$species

group_info <- split(cgMLST_tree$tip.label, f = cgMLST_tree$tip.label)
cgMLST_tree <- groupOTU(cgMLST_tree, group_info)

cgMLST_tree$tip.label <- ids2$id

p_cgmlst <- ggtree(cgMLST_tree, size = 0.2, color = "#808080", layout = "circular")+
  geom_tippoint(aes(color = group), size = 2.5)+
  labs(color = "Species")+
  scale_color_manual(values = palette)+
  guides(color = guide_legend(override.aes = list(size = 8)))+
  theme(legend.position = "right")

palette3 <- c("1-A" = "#8dd3c7",
              "1-I" = "#80b1d3",
              "0" = "grey95")

qr_genes <- mut_report[c("id","gyrA","parC","parE","rpoB","marA","marR")] %>%
  mutate_at(vars(-id),
            funs(ifelse(. == "1", "1-I", "0")))
  
ac_genes <- acquired_report %>%
  mutate_at(vars(-id),
            funs(as.numeric)) %>%
  mutate(qnr = ifelse(rowSums(.[,3:9]) != 0 , 1, 0)) %>%
  select(id, qepA4, qnr) %>%
  mutate_at(vars(-id),
            funs(ifelse(. == "1", "1-A", "0")))

total_gene_data <- qr_genes %>%
  left_join(ac_genes, by = "id") %>%
  column_to_rownames("id")

fan_tree <- open_tree(p_cgmlst, 9)

rotated_fan_tree <- rotate_tree(fan_tree, 95)

final_tree <- gheatmap(rotated_fan_tree,
         total_gene_data,
         offset = 0.005,
         width = 0.2,
         colnames_offset_y = 3,
         colnames_position = "top",
         font.size = 2.5) +
  scale_fill_manual(values = palette3,
                    labels = c("Gene/Mutation absent",
                               "Acquired gene present",
                               "Intrinsic gene mutated"))
final_tree
```


```{r}
# alignment <- Biostrings::readAAMultipleAlignment("../data/alignment.fasta")
# msaR::msaR(alignment)
```

