---
title: "Additional Tables"
author: "HÃ¥kon Kaspersen"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
always_allow_html: yes
output:
    html_document:
      code_folding: hide
      df_print: paged
      highlight: kate
      number_sections: yes
      theme: flatly
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: yes
    pdf_document:
      toc: yes
      toc_depth: '3'
---

# Libraries
```{r load libraries, message=FALSE, warning=FALSE}
# Libraries
library(dplyr)
library(tidyr)
library(kableExtra)
```

```{r Import data, include = FALSE}
# Vector of excluded samples
ex_samples <- c(
  "2016-17-565-1-S",
  "2016-02-415-2-S",
  "2016-02-428-2-S",
  "2016-02-486-2-S",
  "2016-02-732-2-S"
  )

# Collapses data frame to unique lines while ignoring NA's
func_paste <- function(x) paste(unique(x[!is.na(x)]), collapse = ", ")

"%not_in%" <- Negate("%in%")

id_data <- read.table("../data/id.txt",
                      sep = "\t",
                      header = TRUE,
                      stringsAsFactors = F)

isolate_data <- read.table("../data/isolate_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% ex_samples)

two_reads <- read.table("../data/no_of_reads.txt",
                          sep = "\t",
                          header = F,
                          stringsAsFactors = F) %>%
  dplyr::rename("ref" = V1,
         "reads" = V2) %>%
  mutate(ref = gsub(".+/(.*?)\\.fastq\\.gz", "\\1", ref),
         type = ifelse(grepl("R1", ref) == TRUE, "R1", "R2")) %>%
  mutate(ref = gsub("(.*?)_L00[0-9]_R[1-2]_00[0-9]","\\1", ref)) %>%
  group_by(ref) %>%
  mutate(id = n()) %>%
  {. ->> four_reads} %>%
  filter(id == 2) %>%
  spread(type, reads) %>%
  dplyr::rename("R1-1" = R1,
         "R2-1" = R2) %>%
  mutate("R1-2" = NA,
         "R2-2" = NA) %>%
  select(-id) %>%
  mutate_all(funs(as.character(.)))

four_reads <- four_reads %>%
  filter(id == 4) %>%
  group_by(ref) %>%
  mutate(id = 1:n()) %>%
  spread(id, reads) %>%
  summarise_all(funs(func_paste)) %>%
  dplyr::rename("R1-1" = `1`,
         "R2-1" = `2`,
         "R1-2" = `3`,
         "R2-2" = `4`) %>%
  select(-type) %>%
  mutate_all(funs(as.character(.)))

no_of_reads <- as.data.frame(rbind(as.matrix(two_reads),
                                   as.matrix(four_reads))) %>%
  left_join(id_data, by = "ref") %>%
  select(id, everything(), -ref) %>%
  filter(id %not_in% ex_samples)

mic_data <- read.table("../data/mic_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% ex_samples)

mlst_data <- read.table("../data/mlst_results.txt",
                        sep = "\t",
                        header = TRUE,
                        stringsAsFactors = F)
```

```{r message=FALSE}
source("../scripts/ARIBA_analysis.R")

```

# Tables

## Group listing
This table presents the groups created in the isolate selection. The "Additional Resistances" column contains how many additional antimicrobials the isolates in that group is resistant to, other than CIP and/or NAL.
```{r}
mic_data %>%
  group_by(antres, CIP, NAL) %>%
  count() %>%
  dplyr::rename("Additional Resistances" = antres) %>%
  ungroup() %>%
  mutate(Group = 1:n()) %>%
  dplyr::select(Group, everything()) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped","hover"))
```

## Mechanisms in isolates without substitutions in QRDR
This table presents the mechanisms identified in isolates with no amino acid substitutions in either of GyrA, GyrB, ParC and/or ParE.
```{r}
specific_ac_genes <- acquired_report %>%
  gather(PMQR, value, -id) %>%
  filter(value == 1) %>%
  select(-value)

mut_report %>%
  # all gyrB were 0
  filter(gyrA == 0,
         parC == 0,
         parE == 0) %>%
  left_join(specific_ac_genes, by = "id") %>%
  select(-c(nt_change, id)) %>%
  group_by_all() %>%
  dplyr::count() %>%
  ungroup() %>%
  dplyr::select(-c(gyrA,gyrB,parC,parE)) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

## Substitutions in intrinsic genes for isolates with CIP MIC < 0.12
This table presents the substitutions identified in isolates with a MIC-value for ciprofloxacin below the ECOFF. Note that all substitutions in all genes are listed here, not just the QRDR-located ones. 
```{r}
cip_mic_sens <- mic_data %>%
  select(id, CIP, NAL) %>%
  filter(CIP < 0.12)

acquired_genes <- acquired_report %>%
  gather(gene, value, -id) %>%
  filter(value == 1) %>%
  select(-value) %>%
  dplyr::rename("PMQR" = gene)

gene_names <- names(mut_report)
gene_names <- gene_names[-c(1,2)]

mut_table_fixed %>%
  filter(id %in% cip_mic_sens$id,
         gene %in% gene_names) %>%
  select(id, gene, mut) %>%
  spread(gene, mut) %>%
  left_join(acquired_genes, by = "id") %>%
  left_join(mic_data[,c("id", "CIP", "NAL")], by = "id") %>%
  left_join(isolate_data[,c("id","species","method")], by = "id") %>%
  dplyr::select(-c(id, PMQR, method)) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

