---
title: "Additional Tables"
author: "Håkon Kaspersen"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
always_allow_html: yes
output:
    html_document:
      code_folding: hide
      df_print: paged
      highlight: kate
      number_sections: yes
      theme: flatly
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: yes
    pdf_document:
      toc: yes
      toc_depth: '3'
---

```{r message = FALSE, warning=FALSE}
# Libraries
library(dplyr)
library(tidyr)
library(kableExtra)
```

```{r Import data, message=FALSE, warning=FALSE}
# Vector of excluded samples
ex_samples <- c(
  "2016-17-565-1-S",
  "2016-02-415-2-S",
  "2016-02-428-2-S",
  "2016-02-486-2-S",
  "2016-02-732-2-S"
  )

# Collapses data frame to unique lines while ignoring NA's
func_paste <- function(x) paste(unique(x[!is.na(x)]), collapse = ", ")

"%not_in%" <- Negate("%in%")

id_data <- read.table("../data/id.txt",
                      sep = "\t",
                      header = TRUE,
                      stringsAsFactors = F)

isolate_data <- read.table("../data/isolate_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% ex_samples)

mic_data <- read.table("../data/mic_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% ex_samples)

mlst_data <- read.table("../data/mlst_results.txt",
                        sep = "\t",
                        header = TRUE,
                        stringsAsFactors = F)
```

```{r include=FALSE}
source("../scripts/ARIBA_analysis.R")
```

# Group listing
This table presents the groups created in the isolate selection. The "Additional Resistances" column contains how many additional antimicrobials the isolates in that group is resistant to, other than CIP and/or NAL.
```{r}
mic_data %>%
  left_join(isolate_data[,c("id","species")], by = "id") %>%
  group_by(antres, CIP, NAL, species) %>%
  count() %>%
  dplyr::rename("Additional Resistances" = antres) %>%
  ungroup() %>%
  mutate(Group = 1:n()) %>%
  dplyr::select(Group, everything()) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped","hover"))
```

# Mechanisms in isolates without substitutions in QRDR
This table presents the mechanisms identified in isolates with no amino acid substitutions in either of GyrA, GyrB, ParC and/or ParE.
```{r}
specific_ac_genes <- acquired_report %>%
  gather(PMQR, value, -id) %>%
  filter(value == 1) %>%
  select(-value)

mut_report %>%
  # all gyrB were 0
  filter(gyrA == 0,
         parC == 0,
         parE == 0) %>%
  left_join(specific_ac_genes, by = "id") %>%
  select(-c(nt_change, id)) %>%
  group_by_all() %>%
  dplyr::count() %>%
  ungroup() %>%
  dplyr::select(-c(gyrA,gyrB,parC,parE)) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

# Substitutions in intrinsic genes for isolates with CIP MIC < 0.12
This table presents the substitutions identified in isolates with a MIC-value for ciprofloxacin below the ECOFF. Note that all substitutions in all genes are listed here, not just the QRDR-located ones. 
```{r}
cip_mic_sens <- mic_data %>%
  select(id, CIP, NAL) %>%
  filter(CIP < 0.12)

acquired_genes <- acquired_report %>%
  gather(gene, value, -id) %>%
  filter(value == 1) %>%
  select(-value) %>%
  dplyr::rename("PMQR" = gene)

gene_names <- names(mut_report)
gene_names <- gene_names[-c(1,2)]

mut_table_fixed %>%
  filter(id %in% cip_mic_sens$id,
         gene %in% gene_names) %>%
  select(id, gene, mut) %>%
  spread(gene, mut) %>%
  left_join(acquired_genes, by = "id") %>%
  left_join(mic_data[,c("id", "CIP", "NAL")], by = "id") %>%
  left_join(isolate_data[,c("id","species","method")], by = "id") %>%
  dplyr::select(-c(id, PMQR, method)) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```

# Substitutions in GyrB outside QRDR
```{r}
mut_filtered %>%
  filter(gene == "gyrB") %>%
  group_by(mut) %>%
  count() %>%
  filter(mut != "0") %>%
  dplyr::rename("AA Substitution" = mut) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))
```


# R and R-Packages
R version 3.6.0 [1] was used in the present study. The following packages were used:

| Package       | Version | Reference |
|---------------|---------|-----------|
| dplyr         | 0.8.0.1 |   [2]     |
| tidyr         | 0.8.3   |   [3]     |
| ggplot2       | 3.1.1   |   [4]     |
| purrr         | 0.3.2   |   [5]     |
| tibble        | 2.1.1   |   [6]     |
| Biostrings    | 2.50.2  |   [7]     |
| vegan         | 2.5.4   |   [8]     |
| stringr       | 1.4.0   |   [9]     |
| cluster       | 2.0.8   |   [10]    |
| ape           | 5.3     |   [11]    |
| ggtree        | 1.14.6  |   [12]    |
| RColorBrewer  | 1.1-2   |   [13]    |
| cowplot       | 0.9.4   |   [14]    |
| knitr         | 1.22    |   [15]    |
| kableExtra    | 1.1.0   |   [16]    |
| ggsci         | 2.9     |   [17]    |
| fastqcr       | 0.1.2   |   [18]    |
| scales        | 1.0.0   |   [19]    |
| viridis       | 0.5.1   |   [20]    |

# References
[1] RCoreTeam. R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL https://www.R-project.org/. 2018.

[2] Wickham H, François R, Henry L, Müller K. dplyr: A Grammar of Data Manipulation [Internet]. 2019. Available from: https://cran.r-project.org/package=dplyr

[3] Wickham H, Henry L. tidyr: Easily Tidy Data with “spread()” and “gather()” Functions [Internet]. 2019. Available from: https://cran.r-project.org/package=tidyr

[4] Wickham H, Chang W, Henry L, Pedersen TL, Takahashi K, Wilke C, et al. ggplot2: Create Elegant Data Visualisations Using the Grammar of Graphics [Internet]. 2019. Available from: https://cran.r-project.org/package=ggplot2

[5] Henry L, Wickham H. purrr: Functional Programming Tools [Internet]. 2019. Available from: https://cran.r-project.org/package=purrr

[6] Müller K, Wickham H. tibble: Simple Data Frames [Internet]. 2019. Available from: https://cran.r-project.org/package=tibble

[7] Pagès H, Aboyoun P, Gentleman R, DebRoy S. Biostrings: Efficient manipulation of biological strings. 2019.

[8] Oksanen J, Blanchet FG, Friendly M, Kindt R, Legendre P, McGlinn D, et al. vegan: Community Ecology Package [Internet]. 2019. Available from: https://cran.r-project.org/package=vegan

[9] Wickham H. stringr: Simple, Consistent Wrappers for Common String Operations [Internet]. 2019. Available from: https://cran.r-project.org/package=stringr

[10] Maechler M, Rousseeuw P, Struyf A, Hubert M. cluster: “Finding Groups in Data”: Cluster Analysis Extended Rousseeuw et al. [Internet]. 2019. Available from: https://cran.r-project.org/package=cluster

[11] Paradis E, Blomberg S, Bolker B, Brown J, Claude J, Cuong HS, et al. ape: Analyses of Phylogenetics and Evolution [Internet]. 2019. Available from: https://cran.r-project.org/package=ape

[12] Yu G, Smith DK, Zhu H, Guan Y, Lam TT-Y. Ggtree: a package for visualization and annotation of phylogenetic trees with their covariates and other associated data. McInerny G, editor. Methods Ecol Evol [Internet]. 2017 Jan;8(1):28–36. Available from: http://doi.wiley.com/10.1111/2041-210X.12628

[13] Neuwirth E. RColorBrewer: ColorBrewer Palettes [Internet]. 2014. Available from: https://cran.r-project.org/package=RColorBrewer

[14] Wilke CO. cowplot: Streamlined Plot Theme and Plot Annotations for “ggplot2” [Internet]. 2019. Available from: https://cran.r-project.org/package=cowplot

[15] Xie Y. knitr: A General-Purpose Package for Dynamic Report Generation in R [Internet]. 2019. Available from: https://cran.r-project.org/package=knitr

[16] Zhu H. kableExtra: Construct Complex Table with “kable” and Pipe Syntax [Internet]. 2019. Available from: https://cran.r-project.org/package=kableExtra

[17] Xiao N. ggsci: Scientific Journal and Sci-Fi Themed Color Palettes for “ggplot2” [Internet]. 2018. Available from: https://cran.r-project.org/package=ggsci

[18] Kassambara A. fastqcr: Quality Control of Sequencing Data [Internet]. 2019. Available from: https://cran.r-project.org/package=fastqcr

[19] Wickham H. scales: Scale Functions for Visualization [Internet]. 2018. Available from: https://cran.r-project.org/package=scales

[20] Garnier S. viridis: Default Color Maps from “matplotlib” [Internet]. 2018. Available from: https://cran.r-project.org/package=viridis
