---
title: "Clustering analysis and statistics"
author: "Håkon Kaspersen"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
always_allow_html: yes
output:
    html_document:
      code_folding: hide
      df_print: paged
      highlight: kate
      number_sections: yes
      theme: flatly
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: yes
    pdf_document:
      toc: yes
      toc_depth: '3'
---

```{r libraries, message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(ggplot2)
library(ape)
library(tibble)
library(knitr)
library(vegan)
library(distanceR)
library(impoRt)
library(readxl)
library(treeio)
library(adephylo)
library(RColorBrewer)
library(kableExtra)
```


```{r import data}
# Vector of excluded samples
ex_samples <- c(
  "2016-17-565-1-S",
  "2016-02-428-2-S",
  "2016-02-486-2-S",
  "2016-02-732-2-S",
  "2014-01-1741-1-S"
  )

# Collapses data frame to unique lines while ignoring NA's
func_paste <- function(x) paste(unique(x[!is.na(x)]), collapse = ", ")

"%not_in%" <- Negate("%in%")


# Isolate data
id_data <- read.table("../data/id.txt",
                      sep = "\t",
                      header = TRUE,
                      stringsAsFactors = F)

isolate_data <- read.table("../data/isolate_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% ex_samples)

mlst_data <- read.table("../data/mlst_results.txt",
                        sep = "\t",
                        header = TRUE,
                        stringsAsFactors = F)

```

# Minimum spanning tree
The EnteroBase cgMLST scheme [1] for _E. coli_ was used with chewBBACA [2] to generate the cgMLST data.
A minimum spanning tree calculated from the cgMLST data with GrapeTree [3] using MSTreeV2. Number of allele differences are denoted as numbers on branches between nodes. The branches have been collapsed on ten differences or less. The number inside each node is the sequence type based on the EnteroBase 7-gene MLST scheme [4].
```{r, fig.width=15, fig.height=12}
include_graphics("../figures/GrapeTree_cgMLST.png")
```

# Patristic distances
Patristic distances were calculated from the core gene alignment SNP tree.
```{r fig.width=8, fig.height=8, warning=FALSE, message=FALSE}
iqtree <- read.tree("../data/IQTree/core_genes_roary.treefile")
iqtree_clean <- fix_tip_labels(iqtree, id_data, "_pilon_spades.fasta", tree_type = "phylo")

pat_dist <- as.matrix(distTips(iqtree_clean, method = "patristic"))

pat_dist_bin <- as.data.frame(ifelse(pat_dist < 0.04, "1", "0")) %>%
  rownames_to_column("isol1")

row_val <- as.data.frame(pat_dist_bin$isol1) %>%
  rename("id" = `pat_dist_bin$isol1`) %>%
  left_join(mlst_data[,c("id","ST")], by = "id") %>%
  group_by(ST) %>%
  mutate(n = 1:n(),
         new_id = paste(ST, n, sep = "_"))

col_val <- as.data.frame(names(pat_dist_bin)) %>%
  rename("id" = `names(pat_dist_bin)`) %>%
  left_join(mlst_data[,c("id","ST")], by = "id") %>%
  group_by(ST) %>%
  mutate(n = 1:n(),
         new_id = paste(ST, n, sep = "_")) %>%
  ungroup() %>%
  mutate(new_id = ifelse(new_id == "NA_1", "isol1", new_id),
         ST = ifelse(is.na(ST) == TRUE, "isol1", ST))

pat_dist_clean <- pat_dist_bin
names(pat_dist_clean) <- col_val$new_id
pat_dist_clean$isol1 <- row_val$new_id

plot_data <- pat_dist_clean %>%
  gather(isol2, value, -isol1) %>%
  filter(isol1 != isol2)

ggplot(plot_data, aes(reorder(isol1, -as.numeric(value)), reorder(isol2, -as.numeric(value)), fill = value)) +
  geom_tile() +
  scale_fill_manual(values = c("1" = "#80b1d3",
                               "0" = "grey95"),
                    labels = c("Pat. dist > 0.04",
                               "Pat. dist < 0.04")) +
  labs(fill = NULL) +
  theme(axis.text = element_text(size = 3),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.3),
        panel.background = element_blank(),
        axis.ticks = element_blank(),
        axis.title = element_blank()) +
  coord_fixed()
```

# SNP distance statistics on clades of interest
The table presents the summary statistics on the minimum SNP distance to the closest isolate per clade of interest.
```{r}
dist_data <- get_data("../data/Phylogeny/STs",
                      "gubbins_snp_dists.tab",
                      recursive = TRUE,
                      df = FALSE)

names(dist_data) <- c("ST10", "ST117", "ST162", "ST355", "ST58", "ST602", "ST744")

dist_clean <- lapply(dist_data, function(x){
  x %>%
    dplyr::rename("isol1" = `snp-dists 0.6.3`) %>%
    gather(isol2, value, -isol1) %>%
    filter(isol1 != isol2) %>%
    group_by(isol2) %>%
    mutate(min_val = min(value))
})

summary_list <- lapply(dist_clean, function(x) round(summary(x$min_val),0)) %>%
  bind_rows()

summary_list$Type <- c("Min","1st.Q","Median","Mean","3rd.Q","Max")

summary_list %>%
  select(Type, everything()) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"))

```

# Maximum likelihood trees on clades of interest
Each tree is midpoint rooted for visual purposes. The values on each node is the UF Bootstrap values calculated by IQTree. The tip point shape represent the animal species of origin, and the color the year of isolation. The tip labels represent the location of the isolate; the first number represent the county, and the second number the municipality. 
```{r message=FALSE, warning=FALSE}
# Import isolate information and tree metadata
location_data <- read_excel("../data/st_location_data.xlsx") %>%
  select(id, Kommune, Fylke) %>%
  mutate(fyl_id = group_indices(., Fylke),
         kom_id = group_indices(., Kommune)) %>%
  mutate(Location = paste(fyl_id, kom_id, sep = "-")) %>%
  select(id, Location)

tree_metadata <- read.table("../data/chewbbaca/complete_results/tree_metadata.txt",
                            sep = "\t",
                            header = TRUE) %>%
  left_join(location_data, by = "id") %>%
  mutate(Year = substr(id, 1, 4)) %>%
  dplyr::rename("Species" = species)


# Import trees
ST10_raw <- read.iqtree("../data/Phylogeny/STs/ST10/iqtree/iqtree.contree")
ST355_raw <- read.iqtree("../data/Phylogeny/STs/ST355/iqtree/iqtree.contree")
ST602_raw <- read.iqtree("../data/Phylogeny/STs/ST602/iqtree/iqtree.contree")
ST744_raw <- read.iqtree("../data/Phylogeny/STs/ST744/iqtree/iqtree.contree")

raw_trees <- list(ST10_raw,
                  ST355_raw,
                  ST602_raw,
                  ST744_raw)

names(raw_trees) <- c("ST10","ST355","ST602","ST744")

# Clean tree tip labels
clean_trees <- lapply(
  raw_trees,
  function(x) fix_tip_labels(
    x,
    id_data,
    "_pilon_spades.fasta",
    tree_type = "treedata"
    )
  )

# Specify color and shape palettes
palette_shape <- c(
  "Broiler" = 15,
  "Pig" = 18,
  "Red fox" = 16,
  "Wild bird" = 17
)

palette_color <- brewer.pal(11, "Set3")

year_val <- c(
  "2006",
  "2007",
  "2008",
  "2009",
  "2010",
  "2011",
  "2012",
  "2013",
  "2014",
  "2015",
  "2016")

names(palette_color) <- year_val

# Generate trees
annotate_tree(
  clean_trees$ST10,
  tree_metadata,
  layout = "rectangular",
  tree_type = "treedata",
  midroot = TRUE,
  line_width = 0.5,
  label_variable = "Location",
  color_variable = "Year",
  shape_variable = "Species",
  clade_label_node = 19,
  clade_label = "10 SNP diff",
  cladelabel_offset = 0.00055,
  bootstrap_lab = TRUE,
  bootlab_size = 2,
  tippoint_size = 4,
  label_offset = 0.0001,
  shape_palette = palette_shape,
  color_palette = palette_color
) +
  xlim(0, 0.005) +
  ggtitle("ST10 Clade")

annotate_tree(
  clean_trees$ST355,
  tree_metadata,
  layout = "rectangular",
  tree_type = "treedata",
  midroot = TRUE,
  line_width = 0.5,
  label_variable = "Location",
  color_variable = "Year",
  shape_variable = "Species",
  clade_label_node = 26,
  clade_label = "6 SNP diff",
  cladelabel_offset = 0.0013,
  bootstrap_lab = TRUE,
  bootlab_size = 2,
  tippoint_size = 4,
  label_offset = 0.0003,
  shape_palette = palette_shape,
  color_palette = palette_color
) +
  xlim(0, 0.0165) +
  ggtitle("ST355 Clade")

annotate_tree(
  clean_trees$ST602,
  tree_metadata,
  layout = "rectangular",
  tree_type = "treedata",
  midroot = TRUE,
  line_width = 0.5,
  label_variable = "Location",
  color_variable = "Year",
  shape_variable = "Species",
  clade_label_node = 10,
  clade_label = "18 SNP diff",
  cladelabel_offset = 0.001,
  bootstrap_lab = TRUE,
  bootlab_size = 2,
  tippoint_size = 4,
  label_offset = 0.0002,
  shape_palette = palette_shape,
  color_palette = palette_color
) +
  xlim(0, 0.009) +
  ggtitle("ST602 Clade")

annotate_tree(
  clean_trees$ST744,
  tree_metadata,
  layout = "rectangular",
  tree_type = "treedata",
  midroot = TRUE,
  line_width = 0.5,
  label_variable = "Location",
  color_variable = "Year",
  shape_variable = "Species",
  clade_label_node = 12,
  clade_label = "50 SNP diff",
  cladelabel_offset = 0.00024,
  bootstrap_lab = TRUE,
  bootlab_size = 2,
  tippoint_size = 4,
  label_offset = 0.00007,
  shape_palette = palette_shape,
  color_palette = palette_color
) +
  xlim(0, 0.0030) +
  ggtitle("ST744 Clade")

```



# NMDS Analysis of resistance mechanisms
Non-metric multidimensional scaling (NMDS) was used to see if the distribution of quinolone resistance mechanisms from some sequence types were more homogenous than in other sequence types.
```{r}
res_mechanisms <- read.table("../data/MDS_data.txt",
                       sep = "\t",
                       header = TRUE,
                       stringsAsFactors = FALSE) %>%
  column_to_rownames("id")

bin_dist <- dist(res_mechanisms,
                 method = "binary")

nmds_df <- metaMDS(bin_dist,
                   trymax = 200,
                   trace = FALSE) # silence the output

print(paste0("Stress: ", round(nmds_df$stress, 3)))

stressplot(nmds_df)

```


# References
[1] Alikhan N-F et al. A genomic overview of the population structure of Salmonella. Casadesús J, editor. PLOS Genet [Internet]. 2018 Apr 5;14(4):e1007261. Available from: https://dx.plos.org/10.1371/journal.pgen.1007261

[2] Machado MP et al. chewBBACA: A complete suite for gene-by-gene schema creation and strain identification. Microb Genomics. 2018;1–7.

[3] Z Zhou et al. (2018) “GrapeTree: Visualization of core genomic relationships among 100,000 bacterial pathogens”, Genome Res. doi: https://doi.org/10.1101/gr.232397.117

[4] Wirth T et al. (2006) "Sex and virulence in Escherichia coli : an evolutionary perspective", Molecular Microbiology (2006) 60 (5), 1136–1151, doi: 10.1111/j.1365-2958.2006.05172.x