---
title: "Quast Reference Testing"
author: "HÃ¥kon Kaspersen"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
always_allow_html: yes
output:
    html_document:
      code_folding: hide
      df_print: paged
      highlight: kate
      number_sections: yes
      theme: flatly
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: yes
    pdf_document:
      toc: yes
      toc_depth: '3'
---

```{r, include=FALSE}
# Libraries
pacman::p_load(dplyr, ggplot2, tidyr)
```


```{r, include=FALSE}
# Path to folders
path <- "../data/quast_ref_testing"


# Functions
## searches recursively for all filenames with fastqc.zip
file_names_recursive <- function(filepath) {
  files <- list.files(path = filepath, pattern = "transposed_report.tsv", recursive = TRUE)
  return(files)
}

## Creates clean data frame
create_df <- function(df) {
  df <- df %>%
    rename(ctg_0_bp = X..contigs.....0.bp.,
           ctg_1000_bp = X..contigs.....1000.bp.,
           ctg_5000_bp = X..contigs.....5000.bp.,
           ctg_10000_bp = X..contigs.....10000.bp.,
           ctg_25000_bp = X..contigs.....25000.bp.,
           ctg_50000_bp = X..contigs.....50000.bp.,
           tot_len_0_bp = Total.length.....0.bp.,
           tot_len_1000_bp = Total.length.....1000.bp.,
           tot_len_5000_bp = Total.length.....5000.bp.,
           tot_len_10000_bp = Total.length.....10000.bp.,
           tot_len_25000_bp = Total.length.....25000.bp.,
           tot_len_50000_bp = Total.length.....50000.bp.,
           no_of_contigs = X..contigs,
           no_of_misassemblies = X..misassemblies,
           no_of_misassembled_ctg = X..misassembled.contigs,
           no_of_local_misassemblies = X..local.misassemblies,
           unaligned_contigs = X..unaligned.contigs,
           N_per_100kbp = X..N.s.per.100.kbp,
           mismatches_per_100kbp = X..mismatches.per.100.kbp,
           indels_per_100kbp = X..indels.per.100.kbp,
           no_of_genes = X..genes) %>%
    mutate(Assembly = gsub("(.*?)_.+", "\\1", Assembly),
           reference = sub(".+/(.*?)/transposed_report.tsv", "\\1", ref),
           group = sub("(.*?)/.+/transposed_report.tsv", "\\1", ref),
           short_group = ifelse(group == "hiseq_flex_150bp", "A",
                                ifelse(group == "hiseqr_xt_250bp", "B", "C")))
  return(df)
}

func_paste <- function(x) paste(unique(x[!is.na(x)]), collapse = ", ")

```


```{r}
# Data import and wrangling
refs <- read.table("../data/ecoli_ref.tsv", sep = "\t", header = TRUE, stringsAsFactors = FALSE)

filenames <- file_names_recursive(path)

data_list <- lapply(filenames,
                      FUN = function(file) {
                        read.delim(paste0(path, "/", file),
                                   sep = "\t",
                                   header = TRUE,
                                   stringsAsFactors = FALSE)
                      })

data_list <- lapply(data_list, function(df) mutate_all(df, as.character))

names(data_list) <- filenames

df <- bind_rows(data_list, .id = "ref")

df2 <- create_df(df) %>%
  left_join(refs, by = "reference")
```


# Information
## Reference sequences
Below is a table over the references used in the various Quast analyses.
```{r}
refs
```

## Sequencing methodology
```{r}
df2 %>%
  select(group, short_group) %>%
  group_by(short_group) %>%
  summarise_all(funs(func_paste))
```


# Plots
## N50
```{r}
# N50 per reference and sequencing group

ggplot(df2, aes(short_group, as.numeric(N50), fill = short_group))+
  geom_boxplot()+
  labs(y = "N50",
       x = NULL)+
  guides(fill = FALSE)+
  theme_classic()+
  facet_wrap(~name)
```

## Number of contigs
```{r}
# Number of contigs
ggplot(df2, aes(short_group, as.numeric(no_of_contigs), fill = short_group))+
  geom_boxplot()+
  labs(y = "Number of contigs",
       x = NULL)+
  guides(fill = FALSE)+
  theme_classic()+
  facet_wrap(~name)
```

## Number of misassemblies
```{r}
# Number of misassemblies
ggplot(df2, aes(short_group, as.numeric(no_of_misassemblies), fill = short_group))+
  geom_boxplot()+
  labs(y = "Number of misassemblies",
       x = NULL)+
  guides(fill = FALSE)+
  theme_classic()+
  facet_wrap(~name)
```

## Genome fraction
```{r}
ggplot(df2, aes(short_group, as.numeric(Genome.fraction....), fill = short_group))+
  geom_boxplot()+
  labs(y = "Genome fraction (%)",
       x = NULL)+
  guides(fill = FALSE)+
  theme_classic()+
  facet_wrap(~name)
```

## N per 100kbp
```{r}
ggplot(df2, aes(short_group, as.numeric(N_per_100kbp), fill = short_group))+
  geom_boxplot()+
  labs(y = "N per 100 kbp",
       x = NULL)+
  guides(fill = FALSE)+
  theme_classic()+
  facet_wrap(~name)
```

## Largest contigs
```{r}
ggplot(df2, aes(short_group, as.numeric(Largest.contig), fill = short_group))+
  geom_boxplot()+
  labs(y = "Largets Contig Length",
       x = NULL)+
  guides(fill = FALSE)+
  theme_classic()+
  facet_wrap(~name)
```

## Genome fraction vs reference length
```{r}
ggplot(df2, aes(as.numeric(Genome.fraction....), as.numeric(Total.length), fill = short_group))+
  geom_point(pch = 21, size = 3)+
  labs(y = "Total Length",
       x = "Genome Fraction",
       fill = NULL)+
  theme_classic()
```


